Yosemite/Yosemite/Tools/Stats/TopEarnerStatsCachingFetcher.swift

```diff
@@ -0,0 +1,90 @@
+        if let cachedResponse: ProductsReportRequestResponse = try? cache.load(forKey: cacheKey) {
+            response = cachedResponse.items
+        } else {
+            response = try await productsReportsRemote.loadTopProductsReport(for: siteID,
+                                                                                       timeZone: timeZone,
+                                                                                       earliestDateToInclude: earliestDateToInclude,
+                                                                                       latestDateToInclude: latestDateToInclude,
+                                                                                       quantity: quantity)
+            let cacheEntry = CodableCacheEntry(value: ProductsReportRequestResponse(items: response), timeToLive: Constants.cacheTimeToLive)
+            cache.save(cacheEntry, forKey: cacheKey)
+        }
+
+        return convertProductsReportIntoTopEarners(siteID: siteID,
+                                                   granularity: timeRange.topEarnerStatsGranularity,
+                                                   date: latestDateToInclude,
+                                                   productsReport: response,
+                                                   quantity: quantity)
+    }
+}
+
+private extension TopEarnerStatsCachingFetcher {
+    /// Converts the `[ProductsReportItem]` list in a Products analytics report into `TopEarnerStats`
+    ///
+    func convertProductsReportIntoTopEarners(siteID: Int64,
+                                             granularity: StatGranularity,
+                                             date: Date,
+                                             productsReport: [ProductsReportItem],
+                                             quantity: Int) -> TopEarnerStats {
+        let statsDate = StatsStoreV4.buildDateString(from: date, with: granularity)
+        let statsItems = productsReport.map { product in
+            TopEarnerStatsItem(productID: product.productID,
+                               productName: product.productName,
+                               quantity: product.quantity,
+                               total: product.total,
+                               currency: "", // TODO: Remove currency https://github.com/woocommerce/woocommerce-ios/issues/2549
+                               imageUrl: product.imageUrl)
+        }
+        return TopEarnerStats(siteID: siteID, date: statsDate, granularity: granularity, limit: quantity.description, items: statsItems)
+    }
+
+    func requestCacheKey(siteID: Int64,
+                         timeZone: TimeZone,
+                         earliestDateToInclude: Date,
+                         latestDateToInclude: Date,
+                         quantity: Int) -> String {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "dd:MM:yy"
+
+        let stringArray = [String(siteID),
```

Comment:
At the beginning I only consider the date ranges to cache, but then I believe that the quantity and the time zone (and of course the site id) should be taken into account. @rachelmcr What do you think?

Reply:
Given that the quantity and time zone are also parameters for the request, it makes sense to include them. That said, a couple thoughts come to mind:

* We might want to take this opportunity to enforce using the site time zone (not device time zone) for stats requests, for consistency throughout the app.
* If we make the same request in multiple places in the app, but for different quantities, are we wasting disk space / losing an opportunity to use what's already cached? I think right now we're always requesting a quantity of 5, so for practical purposes right now I don't think it matters either way. But what if we request a higher quantity in one part of the app and cache that, and then request fewer elsewhere and don't use the cached data because the quantity is different?